### PR TITLE
Add YouTube + Slides Demo

### PR DESCRIPTION
**Pull request by grant**
I use this demo in my "All About Apps Script Talks". I'd like to add it to GitHub so the sample doesn't get lost in my Drive files.

I can additionally create a quickstart/article with this demo if needbe.

### CODE REVIEW

**Comment by grant** [comment]
Besides the 2 Travis lint failures, PTAL.

File: advanced/youtube.gs [hunk 1]
```diff
@@ -100,3 +100,54 @@ function addSubscription() {
...
+ * Logs out the URL of the presentation.
+ */
+function createSlides() {
+  var youTubeVideos = getYouTubeVideosJSON(YOUTUBE_QUERY);
+  var presentation = SlidesApp.create(PRESENTATION_TITLE);
+  presentation.getSlides()[0].getPageElements()[0].asShape().getText().setText(PRESENTATION_TITLE);
+  
+  // Add slides with videos and log the presentation URL to the user.
+  for (var i = 0; i < youTubeVideos.length; ++i) {
+    if (youTubeVideos[i].url.indexOf('undefined') === -1) {
```
[the remark this answers is not in the document]
Comment (grant) [reply advanced/youtube.gs]: I believe some videos are private but still show up in the API with limited data.
I've moved this filter upstream as suggested 👍

File: advanced/youtube.gs [hunk 2]
```diff
@@ -100,3 +100,54 @@ function addSubscription() {
...
+      var slide = presentation.appendSlide();
+      slide.insertVideo(youTubeVideos[i].url, 0, 0, presentation.getPageWidth(), presentation.getPageHeight());
```
[the remark this answers is not in the document]
Comment (grant) [reply advanced/youtube.gs]: Fixed.

File: advanced/youtube.gs [hunk 3]
```diff
@@ -100,3 +100,54 @@ function addSubscription() {
...
+ * @ref https://developers.google.com/youtube/v3/docs/search/list
+ */
+function getYouTubeVideosJSON(query) {
+  var youTubeResults = YouTube.Search.list('id,snippet', { 
+    q: query,
+    maxResults: 10
+  });
+
+  var json = [];
+  for (var i = 0; i < youTubeResults.items.length; i++) {
```
[the remark this answers is not in the document]
Comment (erickoledadevrel) [reply advanced/youtube.gs]: You still appear to be using a for loop here.

File: advanced/youtube.gs [hunk 4]
```diff
@@ -26,10 +26,9 @@ function searchByKeyword() {
     maxResults: 25
   });
 
-  for (var i = 0; i < results.items.length; i++) {
-    var item = results.items[i];
+  results.items.map(function(item) {
```
Comment (erickoledadevrel) [inline advanced/youtube.gs]: Optional: I prefer forEach() when I don't intend to return anything.

File: advanced/youtube.gs [hunk 5]
```diff
@@ -112,22 +111,25 @@ var YOUTUBE_QUERY = 'San Francisco, CA';
...
  * Gets a list of YouTube videos.
  * @param {String} query - The query term to search for.
+ * @return {object[]} A list of objects with YouTube video data.
  * @ref https://developers.google.com/youtube/v3/docs/search/list
  */
 function getYouTubeVideosJSON(query) {
-  var youTubeResults = YouTube.Search.list('id,snippet', { 
+  var youTubeResults = YouTube.Search.list('id,snippet', {
     q: query,
     maxResults: 10
```
Comment (erickoledadevrel) [inline advanced/youtube.gs]: Add type: 'video', to ensure you aren't returning playlists and channels.

File: advanced/youtube.gs [hunk 6]
```diff
@@ -112,22 +111,25 @@ var YOUTUBE_QUERY = 'San Francisco, CA';
...
 
   var json = [];
   for (var i = 0; i < youTubeResults.items.length; i++) {
     var item = youTubeResults.items[i];
-    json[i] = {
-      url: 'https://youtu.be/' + item.id.videoId,
-      title: item.snippet.title,
-      thumbnailUrl: item.snippet.thumbnails.high.url
-    };
+    if (item.id.videoId) { // Some videos are private and don't have a videoId.
```
Comment (erickoledadevrel) [inline advanced/youtube.gs]: I find it surprising that the search is returning private videos. Can you confirm that is the case? Or is it returning channels and playlists, as mentioned above.

File: advanced/youtube.gs (hunk 4 above)
Comment (grant) [reply advanced/youtube.gs]: Sure. Changed.

File: advanced/youtube.gs (hunk 5 above)
Comment (grant) [reply advanced/youtube.gs]: Great. This should fix the previous issue.

File: advanced/youtube.gs (hunk 6 above)
Comment (grant) [reply advanced/youtube.gs]: I've fixed this with the `type` param.

**Comment by grant** [comment]
> You still appear to be using a for loop here.

I removed for-i loops.
I took optional comments as optional. I ended up using the forEach loop.

We're going to eventually use V8 for-of loops, which uses native JS loops –rather than creating a new function closure with the forEach `function` prototype. As such, `forEach` is not used in modern JS and I prefer not using them.

The diff will be a bit bigger if we convert every for loop to `forEach` rather than for-i/for-of.

**Comment by grant** [comment]
@erickoledadevrel PTAL.

**Pull request merged**